### PR TITLE
"Compiler"

### PR DESCRIPTION
**Pull request by flamion**

### CODE REVIEW

File: backend/shork/src/main/kotlin/internal/compiler/Parser.kt [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+
+import software.shonk.interpreter.internal.addressing.AddressMode
+import software.shonk.interpreter.internal.addressing.Modifier
+import software.shonk.interpreter.internal.instruction.*
+import software.shonk.interpreter.internal.instruction.AbstractInstruction
+import software.shonk.interpreter.internal.instruction.Dat
+import software.shonk.interpreter.internal.instruction.Mov
+import software.shonk.interpreter.internal.instruction.Split
+
+internal class Parser(private val tokens: List<Token>) {
```
Comment (PietHelzel) [inline backend/shork/src/main/kotlin/internal/compiler/Parser.kt:11]: Describe briefly what the Parser actually does

File: backend/shork/src/main/kotlin/internal/compiler/Parser.kt [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+    private var current = 0
+    private var instructions: MutableList<AbstractInstruction> = ArrayList()
+    var errors: MutableList<Pair<String, Token>> = ArrayList()
```
Comment (PietHelzel) [inline backend/shork/src/main/kotlin/internal/compiler/Parser.kt:14]: What does this error list contain? (What are the String and Token?)

File: backend/shork/src/main/kotlin/internal/compiler/Parser.kt (hunk 1 above)
Comment (PietHelzel) [inline backend/shork/src/main/kotlin/internal/compiler/Parser.kt:11]: Describe briefly what the Parser actually does

File: backend/shork/src/main/kotlin/internal/compiler/Parser.kt [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+                null
+            }
+            else -> {
+                emitError("Unexpected token, expected an instruction", token)
+                null
+            }
+        }
+    }
+
+    // Grab the modifier, if it exists
```
Comment (PietHelzel) [inline backend/shork/src/main/kotlin/internal/compiler/Parser.kt:194]: Maybe make a javadoc comment out of this

File: backend/shork/src/main/kotlin/internal/compiler/Scanner.kt [hunk 4]
```diff
@@ -0,0 +1,173 @@
+package software.shonk.interpreter.internal.compiler
+
+import java.util.*
+import kotlin.collections.ArrayList
+
+internal class Scanner(private val source: String) {
```
Comment (PietHelzel) [inline backend/shork/src/main/kotlin/internal/compiler/Scanner.kt:6]: Briefly describe the purpose of the Scanner

File: backend/shork/src/main/kotlin/internal/compiler/Scanner.kt [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    private val tokens: MutableList<Token> = ArrayList()
+    private var start = 0
+    private var current = 0
+    private var line = 1
+    private val keywordMap: HashMap<String, TokenType> = HashMap<String, TokenType>()
+    val errors: MutableList<Pair<Int, String>> = ArrayList()
```
Comment (PietHelzel) [inline backend/shork/src/main/kotlin/internal/compiler/Scanner.kt:12]: What does this error list actually contain? (What are the Int and String?)

File: backend/shork/src/main/kotlin/internal/compiler/Scanner.kt [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+            return '\u0000'
+        }
+        return source[current]
+    }
+
+    /**
+     * Used for lexemes like `++`, `--`, `==`, `!=`, `<=`, `>=`.
+     *
+     * Basically a conditional advance()
+     */
```
Comment (PietHelzel) [inline backend/shork/src/main/kotlin/internal/compiler/Scanner.kt:113]: Make this comment a bit clearer, describe what the match function actually does.

File: backend/shork/src/main/kotlin/internal/compiler/Token.kt [hunk 7]
```diff
@@ -0,0 +1,29 @@
+package software.shonk.interpreter.internal.compiler
+
+internal class Token(val type: TokenType, val lexeme: String, val literal: Any, val line: Int) {
```
Comment (PietHelzel) [inline backend/shork/src/main/kotlin/internal/compiler/Token.kt:3]: Describe clearly what data this Token class contains. What is the lexeme and especially what is the literal?

File: backend/shork/src/test/kotlin/compiler/TestParser.kt [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+                            listOf(instance),
+                        )
+                    )
+                }
+            }
+
+            return arguments
+        }
+
+        private fun generateAlwaysModifierXXX(): List<Arguments> {
```
Comment (PietHelzel) [inline backend/shork/src/test/kotlin/compiler/TestParser.kt:305]: This needs more explanation, as the function name isn't clear enough

**Review by PietHelzel — CHANGES_REQUESTED** [review]

File: backend/shork/src/main/kotlin/internal/compiler/Token.kt (hunk 7 above)
Comment (flamion) [reply backend/shork/src/main/kotlin/internal/compiler/Token.kt:3]: Explaining domain knowledge should not be the purpose of a comment, unless maybe if it is very specific

File: backend/shork/src/test/kotlin/compiler/TestParser.kt (hunk 8 above)
Comment (flamion) [reply backend/shork/src/test/kotlin/compiler/TestParser.kt:305]: Already addressed :)

File: backend/shork/src/main/kotlin/internal/compiler/ExperimentWithParsing.kt
Comment (nikkischnelle) [inline backend/shork/src/main/kotlin/internal/compiler/ExperimentWithParsing.kt:1]: why is this still present?

**Review by PietHelzel — DISMISSED** [review]
lgtm

File: backend/shork/src/main/kotlin/internal/compiler/ExperimentWithParsing.kt
Comment (flamion) [reply backend/shork/src/main/kotlin/internal/compiler/ExperimentWithParsing.kt:1]: You're right, this was missed

**Review by nikkischnelle — APPROVED** [review]
lgtm

**Review by PietHelzel — APPROVED** [review]
please add a few more lines to reach 1420 lines again >:(

**Pull request closed without merging**